### PR TITLE
Improve parser issue detection

### PR DESCRIPTION
**Pull request by albrechtd**
This patch aims at improving the detection of issues in the GMime parser:

(1) Illegal header values
Catch more cases where the parser detects a malformed header name, and always report it as critical issue as the parser may skip the entire message or parts of it.

(2) Detect more duplicated headers
Detect duplicated headers which may exist only once according to RFC 5322, Sect. 3.6.  Whilst a repetition of the same value can be used as spam indicator, a different value could be used to break DKIM signatures (see <http://noxxi.de/research/breaking-dkim-on-purpose-and-by-chance.html#hdr2>).

(3) Report RFC 2047 violation
A space is not allowed within a RFC 2047 encoded-word and may be a spam indicator.

Note that some parser warning codes have been renamed as to describe their purpose clearer.

Details:
========
* gmime/gmime-internal.h: add offset parameter to _g_mime_utils_header_decode_(text|phrase); new function _internet_address_list_parse
* gmime/gmime-header.c, gmime/gmime-message.c: use modified/new internal functions
* gmime/gmime-param.c: pass offset to decode_param
* gmime/gmime-parser-options.h: rename some warning codes, broken header name is always critical, add RFC 2047 error
* gmime/gmime-parser.c: replace is_7bit_clean by the functionally equivalent g_mime_utils_text_is_8bit; always warn about broken header name; refactor and add code for detecting more duplicated headers; properly set parser options when scanning an embedded message part
* gmime/gmime-utils.c: detect SP in RFC 2047 encoded-word; pass offset to _g_mime_utils_header_decode_(text|phrase)
* gmime/internet-address.c: pass offset to decode_name, address_list_parse, group_parse, address_parse; add new function _internet_address_list_parse with an additional offset parameter, and make internet_address_list_parse a wrapper for it
* examples/msgcheck.c: use new warning codes in the example application

### CODE REVIEW

File: gmime/gmime-parser.c [hunk 1]
```diff
@@ -922,7 +909,7 @@ header_parse (GMimeParser *parser, GMimeParserOptions *options)
 		priv->header_cb (parser, header->name, header->raw_value,
 				 header->offset, priv->user_data);
 	
-	if (can_warn && (!is_7bit_clean (header->name) || !g_utf8_validate (header->raw_value, -1, NULL)))
+	if (can_warn && (g_mime_utils_text_is_8bit ((const unsigned char *) header->name, strlen (header->name)) || !g_utf8_validate (header->raw_value, -1, NULL)))
```
Comment (jstedfast) [inline gmime/gmime-parser.c]: Any reason you chose to use g_mime_utils_text_is_8bit() instead of is_7bit_clean() other than attempting to reduce code duplication?

One problem with the is_8bit() function is that you have to strlen() the field name whereas with is_7bit_clean(), you don't. This is just a performance consideration and most likely minor due to the fact that field names are typically short, but I figured I'd bring it up anyway.

File: gmime/gmime-parser.c [hunk 2]
```diff
@@ -975,6 +962,18 @@ has_content_headers (GPtrArray *headers)
...
+		gchar *eol; 													\
+		gchar *bad_header; 												\
+		 	 	 	 	 	 	 	 	 	 	 				\
+		for (eol = inptr; eol < inend && eol[0] != '\r' && eol[0] != '\n'; eol++); 					\
+		bad_header = g_strndup (start, eol - start); 									\
+		_g_mime_parser_options_warn (options, priv->header_offset, GMIME_CRIT_INVALID_HEADER_NAME, bad_header);		\
+		g_free (bad_header); 												\
+	} 															\
+} G_STMT_END
+
```
Comment (jstedfast) [inline gmime/gmime-parser.c]: I'd probably prefer that this be a function rather than a macro.

File: gmime/gmime-utils.c [hunk 3]
```diff
@@ -1468,7 +1469,9 @@ rfc2047_token_new_encoded_word (const char *word, size_t len)
 	token = rfc2047_token_new (payload, inptr - payload);
 	token->charset = g_mime_charset_iconv_name (charset);
 	token->encoding = encoding;
-	
+	/* RFC 2047 forbids SP in the encoded-word */
+	token->sp_in_encword = memchr(token->text, ' ', token->length) ? 1 : 0;
+
```
Comment (jstedfast) [inline gmime/gmime-utils.c]: It forbids more than just spaces... there's a list of special characters that are not allowed (and it also depends on whether you are encoding a `phrase` vs unstructured text).

The `tokenize_rfc2047_phrase` and `tokenize_rfc2047_text` functions have STRICT vs LOOSE branches. The STRICT branch is already strict enough that it won't allow spaces in an encoded-word token.

The LOOSE branch allows pretty much any slop that looks vaguely like an encoded-word token.

A simpler/less expensive way of doing this would just be to have those methods keep an extra bool state variable to say whether it is "valid" or not and then only warn if `if ((token = rfc2047_token_new_encoded_word (word, n))) {` evaluates to true.

I would probably also just warn in the tokenize functions rather than tokenizing and then scanning the tokens to see if any of them have sp_in_encword set.

In other words...

```c
if ((token = rfc2047_token_new_encoded_word (word, n))) {
    if (has_specials)
        _g_mime_parser_options_warn (options, offset, GMIME_WARN_INVALID_RFC2047_HEADER_VALUE, text);

    ...
```

Comment (albrechtd) [reply gmime/gmime-utils.c]: Yes, that's a much better approach.  Will fix that, too…

BTW, I think it's about time to send you a big __thank you__ for your great work and this extremely useful library!  And of course I wish you a happy Christmas and all the best for the new year!

File: gmime/gmime-parser.c (hunk 2 above)
Comment (albrechtd) [reply gmime/gmime-parser.c]: I agree that the macro is ugly.  As a function would require 5 (or 6, if we pass can_warn) parameters, it seemed to be easier to me, though.  I will fix that…

File: gmime/gmime-parser.c (hunk 1 above)
Comment (albrechtd) [reply gmime/gmime-parser.c]: Yes, the reason was an attempt to remove duplicated code.  IMO, the additional strlen() will typically not introduce any overhead, as the “usual” case will be can_warn == FALSE, i.e. strlen() is not called.

File: gmime/gmime-utils.c (hunk 3 above)
Comment (jstedfast) [reply gmime/gmime-utils.c]: Don't worry about it, I've already merged your patch and made the adjustments myself.

And thanks! :)

Merry Christmas and a happy New Year to you as well!

File: gmime/gmime-parser.c (hunk 1 above)
Comment (jstedfast) [reply gmime/gmime-parser.c]: That's fine. I merged that as it is and left it alone. I just wanted to be aware of any problems if there were any :)
